ledger/participant-state/kvutils/src/main/protobuf/daml_kvutils.proto

```diff
@@ -392,19 +465,73 @@ message DamlStateValue {
   }
 }
 
+// These keys identify intents to change the Daml ledger using a Daml
+// transaction built from a DamlCommand.
 message DamlCommandDedupKey {
+  // Intents are namespaced to the list of submitting parties.
   repeated string submitters = 1;
-  reserved 2; // was application_id
+
+  // Submitter provided application identifiers.
+  // Optional. Provided for rank-based deduplication.
+  reserved string application_id = 2;
+
+  // An identifier for the intent.
   string command_id = 3;
 }
 
+// The DamlCommand deduplication logic ensures that Ledger API clients can
+// implement exactly-once submission of DamlCommands using retries.
+// The logic, called rank-based deduplication, works on a
+// per-DamlCommandDedupKey basis. It stores the rank and result of the last
+// submission that was validated, and cancels any further submissions with the
+// same rank instead of executing them.
+//
+// DamlCommandDedupValues are tagged with an expiry date to allow ledgers to
+// reclaim storage space by deleting them. The expiry date is provided either
+// directly or indirectly by Ledger API clients. They are expected to provide
+// a close, but not too close expiry data, so that ledger can reclaim storage
+// space timely without the clients experiencing duplicate submissions.
 message DamlCommandDedupValue {
   reserved 1; // was record_time
-  // the time until when future commands with the same
-  // deduplication key will be rejected due to a duplicate submission
+
+  // The time until which ledgers MUST keep this DamlCommandDedupValue around.
+  // Ledgers MAY delete them after this time measured against their local
+  // clock.
+  // Note: we exploit that Ledger API clients are advised to not reuse
+  // DamlCommandDedupKeys, as they cannot assume that their timely removal.
   google.protobuf.Timestamp deduplicated_until = 2;
+
+  // The identity of the submission whose rank and result is stored.
+  // This identity is typically a UUID provided by Ledger API clients.
```

Comment:
I'd remove the UUID reference here as it's an opaque string for KV ledgers. How about "ledger string"?

Reply:
Ack.